### PR TITLE
Use FOR KEY SHARE UPDATE instead of FOR UPDATE

### PR DESCRIPTION
**Pull request by dstandish**
Pretty sure we are taking a stronger lock than we need to.

Why this matters:

Taking the weaker lock FOR KEY SHARE UPDATE allows other tables to do concurrent inserts with fks to the locked row.   This matters e.g. when we insert a task reschedule that keys to dag run.  

Right now we wait (before inserting reschedule) to be able to lock the row on dag run.  But if you have many failures close in time this can be trouble.  And it's unnecessary since we're just inserting a row with a fk to dag run.  But logic was added to fully lock the dag run because there were reported deadlocks.  But if we use the weaker lock in the first place, that would not be an issue.

And since, as far as i know, we're not ever going to change the key of the locked row, it should always be safe to do this.

Here's some relevant docs https://www.postgresql.org/docs/current/explicit-locking.html#LOCKING-ROWS

And this table shows which types of locks conflict with which.

image

Interactive testing notes:
* ran a few dags with 2 schedulers.  seemed to work fine.

cc @seanmuth @nsAstro

### CODE REVIEW

File: airflow/utils/sqlalchemy.py [hunk 1]
```diff
@@ -341,7 +341,7 @@ def with_row_locks(
         kwargs["nowait"] = True
     if skip_locked:
         kwargs["skip_locked"] = True
-    return query.with_for_update(**kwargs)
+    return query.with_for_update(**kwargs, key_share=True)
```
Comment (uranusjr) [inline airflow/utils/sqlalchemy.py]: Should this be configurable in the function? (i.e. add this as an argument and defaulting to `True`)

File: airflow/models/taskinstance.py [hunk 2]
```diff
@@ -1751,15 +1749,6 @@ def _handle_reschedule(
...
 
-    # Lock DAG run to be sure not to get into a deadlock situation when trying to insert
-    # TaskReschedule which apparently also creates lock on corresponding DagRun entity
-    with_row_locks(
-        session.query(DagRun).filter_by(
-            dag_id=ti.dag_id,
-            run_id=ti.run_id,
-        ),
-        session=session,
-    ).one()
```
Comment (uranusjr) [inline airflow/models/taskinstance.py:1771]: Is this related? (I’m not saying I understand what the code does to begin with)

**Comment by seanmuth** [comment]
Can provide more details tomorrow, but I was able to reproduce an issue seen at Astronomer due to this aggressive locking.

With a DAG with 1000 FileSensors set to `reschedule` mode, and 10 tries. Launching 5 runs of this DAG tanks the scheduler's heartbeat, as soon as the sensor tasks begin rescheduling. SQL statements back up, all waiting for tuple locks on the same `dag_run` record. The scheduler then also gets stuck waiting behind these lock statements, and it starts missing heartbeats.

Planning to patch this PR in tomorrow morning to test its effects.

File: airflow/models/taskinstance.py (hunk 2 above)
Comment (dstandish) [reply airflow/models/taskinstance.py:1771]: Yes it is.  By doing `for key share update` that means we no longer block `select for key share` which means this insert (few lines below) will not conflict with dag run locks.   so there's no reason to lock the dag run!

File: airflow/utils/sqlalchemy.py (hunk 1 above)
Comment (dstandish) [reply airflow/utils/sqlalchemy.py]: it's not really needed at this time AFAIK but yeah that would be a reasonable thing to do :)

File: airflow/models/taskinstance.py (hunk 2 above)
Comment (dstandish) [reply airflow/models/taskinstance.py:1771]: removing this lock is really the main point of this PR.  using the weaker lock in scheduler is in service of this.

File: airflow/utils/sqlalchemy.py (hunk 1 above)
Comment (dstandish) [reply airflow/utils/sqlalchemy.py]: done btw

File: airflow/models/taskinstance.py [hunk 3]
```diff
@@ -1760,16 +1758,16 @@ def _handle_reschedule(
...
-    # Log reschedule request
+    # set state
+    ti.state = TaskInstanceState.UP_FOR_RESCHEDULE
+
+    ti.clear_next_method_args()
+
+    session.merge(ti)
+    session.commit()
```
Comment (dstandish) [inline airflow/models/taskinstance.py:1767]: @potiuk if anyone would have concerns with this, i suspect it would be you.  will also tag @jscheffl for good measure.

**Comment by kaxil** [comment]
> Can provide more details tomorrow, but I was able to reproduce an issue seen at Astronomer due to this aggressive locking.
> 
> With a DAG with 1000 FileSensors set to `reschedule` mode, and 10 tries. Launching 5 runs of this DAG tanks the scheduler's heartbeat, as soon as the sensor tasks begin rescheduling. SQL statements back up, all waiting for tuple locks on the same `dag_run` record. The scheduler then also gets stuck waiting behind these lock statements, and it starts missing heartbeats.
> 
> Planning to patch this PR in tomorrow morning to test its effects.

@seanmuth Any update on the results of this patch?

File: airflow/models/taskinstance.py (hunk 3 above)
Comment (potiuk) [reply airflow/models/taskinstance.py:1767]: The only concern I have that it looks like Airflow 2 change only. The interface there will change and rescheduling should be handled in the "API" component, not in the worker, so this code does not really matter and will be heavily changed when task sdk is implemented. Am I right @ashb ?

Comment (potiuk) [reply airflow/models/taskinstance.py:1767]: Some more explanation:

Generally with Airflow 3 the "transaction" boundaries for any operation on the database will have to change heavily - because basically when you reach out to the API component, there is no way to make several different calls in the same transaction - so the transaction boundary will be necessarily smaller than every API call implemented. And there will be possibly more transactions within the same call. 

This was one of the major issues with the "internal-api" approach that we tried to maintain the same transaction boundaries "with DB" and "without DB" and it led to some spaghetti code - where we had to move related DB code from one transaction together in a single `@internal_api` call.

Here, I am not sure how "reschedule" will be converted to the new API in Airflow 3 - and where transaction boundaries will be, but in any case I think this code will be looking quite differently, I think.

Comment (dstandish) [reply airflow/models/taskinstance.py:1767]: Here's my thought: we don't need the stronger lock, so better to use the weaker lock.

It does not only affect reschedule, but any insert on a table that has an FK to the locked table.  We use this somewhat broadly.

Comment (potiuk) [reply airflow/models/taskinstance.py:1767]: I recalled something. I vaguely recal there was some problem with mysql that FOR UPDATE was actually locking the FKs connected - and it caused deadlocks - because those indexes could be locked in other transactions in different sequence. I think this is the main reason why we have this strong lock here.

But if we make the change for postgres only, then I guess that one is not a problem.

Comment (potiuk) [reply airflow/models/taskinstance.py:1767]: No concerns because of that.

Comment (dstandish) [reply airflow/models/taskinstance.py:1767]: @potiuk yes, you are correct, this lock was added to address a report of deadlock _(added by you, incidentally :) )_. and i think user was mysql.   but it would have worked the same way on postgres.  postgres FKs work the same way.  when you insert it has to lock the row with the PK that you are referencing so you don't end up with a reference to a row that does not exist.

postgres is unique in that it provides a way to say "i'm not going to delete or change the pk of this row" which means the insert can go through knowing the row won't disappear.  that's the purpose of this lock mode.

but anyway, so yes in a sense this PR is "postgres only" in that we only change the lock used for concurrency management.  but since mysql doesn't have that mode, i had to still ensure we protect against deadlock in mysql.  and the way i did that was to introduce a commit immediately before inserting taskreschedule.  this will avoid deadlock in mysql because there will not be any other operation in the transaction where the TR record is inserted.  and since there will be nothing else in the tran, then a deadlock will be impossible.  i do not see any downside to committing before.  this means that with mysql, in contrast with postgres, it will still need to wait until there's nothing taking a FOR UPDATE lock on the record (postgres won't have to wait), but at least it won't _also_ have deadlock risk in addition to waiting; and it won't have to also do its own FOR UPDATE lock of the TI.

Comment (potiuk) [reply airflow/models/taskinstance.py:1767]: The only potential drawback is that if the system crashes right after commit, the task status will be updated, but TaskReschedule will not be created. Likelihood of it is small, but I think it's worth to at least do a mental exercise of what happens then and whether it's worth it.

Now, when I think of it, it's little worrying - because it actually happens in "_run_raw_task" - which means that it has **some** probabilty to happen (OOM and others). 

BTW. This part should look pretty differently for Airflow 3 with AIP-72 because reschedule will not happen in task but in the API component most likely, so I am not quite sure if that is targetted to be only back-ported? 

@ashb - any concerns and comments here?

Comment (dstandish) [reply airflow/models/taskinstance.py:1767]: Yeah I thought about that.  To me it seems a good tradeoff.  Let's discuss.

First, the probability would seem to be quite low. Because we're talking about a very small time window. 

```
1. session.commit
2. session.add
3. session.commit
```
the time is between 1 and 3. 

If 2 is stuck in lock wait that could become less insignifigant (i.e. more frequent).

but i think either way it seems we arrive at a better outcome.

why?

because let's imagine the worker blows up between 1 and 3.   What is the consequence?

I believe the consequence would be that no TR record is inserted although the state is set properly.  

That scenario looks to be handled here in the relevant dep: https://github.com/apache/airflow/blob/85fd0e1102d664337d6bb08e590979867372f61d/airflow/ti_deps/deps/ready_to_reschedule.py#L85

So the sensor would be resumed probably more quickly than it thought it would.

But with the before code, it's actually worse what happens.  With the old code if we imagine worker blows up at that same time, then the task does not report state correctly, it would just be an OOM task death which would be a zombie that would be detected by scheduler and the sensor would not resume poking until after retries if configured.

So commiting there that seems better to me anyway, if you assume that it blows up immediately after.

And, commiting there means no deadlock concern, and it means we don't need to hold the table with a select for update, which holds the table for significantly longer, and if you have lots of other sensors in same dag, it can cause trouble.

Comment (dstandish) [reply airflow/models/taskinstance.py:1767]: One other nuance here.

Before, we wait to take an exclusive lock on dag run with select for update.

And all other tasks that are trying to write to TaskReschedule must do the same.  Thus if there are multiple sensors in same dag, they will block each other and pile up.  But if we remove the dag run lock, then the only lock obtained is the implicit SELECT FOR SHARE taken internally by mysql when it inserts the record into TR.  And this does not block other sensors from taking the same lock.

So to me, this would just have a positive effect on throughput and not really any meaningful downside.

Comment (potiuk) [reply airflow/models/taskinstance.py:1767]: Quite agree. No concerns any more.

**Comment by prabhusneha** [comment]
The changes LGTM. 
I don't see any issues for MySQL.

**Review by potiuk — APPROVED** [review]

**Review by eladkal — APPROVED** [review]
